### ISSUE TITLE
Astro.request.headers warnings during build with checkOrigin enabled

### ISSUE THREAD

**Issue by jurajkapsz** [withastro/astro#12152]
### Astro Info

```block
Astro                    v4.15.12
Node                     v20.11.1
System                   Linux (x64)
Package Manager          pnpm
Output                   hybrid
Adapter                  @astrojs/node
Integrations             @astrojs/mdx
                         @astrojs/sitemap
```

### Describe the Bug

Confirming [this comment](https://github.com/withastro/astro/issues/12069#issuecomment-2374407210), with astro config option: 
```js
  security: {
    checkOrigin: true,
  },
```
there is a console warning:
```txt
[WARN] `Astro.request.headers` is unavailable in "static" output mode, and in prerendered pages within "hybrid" and "server" output modes. If you need access to request headers, make sure that `output` is configured as either `"server"` or `output: "hybrid"` in your config file, and that the page accessing the headers is rendered on-demand.
```
shown on every built `html` page during build.

I asume I am accessing `Astro.request.headers` through `Middleware`, for static build purposes.
```js
export const onRequest = defineMiddleware(({ locals, request }, next) => {
...
});
```
My site is all static, and I use Astro Actions to handle forms, which [actions] I call from a `<script>` tag on page, so I use the `hybrid` mode.

### What's the expected result?

This warning was maybe fixed in the past but is still present with the `checkOrigin` config option, in combination with static mode pages that access `Astro.request.header` during their build time.

### Link to Minimal Reproducible Example

#

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by ematipico** [comment]
I am not sure I understand why I enable a feature that isn't meant for SSG 🤔 Please provide a reproduction.

**Comment by jurajkapsz** [comment]
> I am not sure I understand why I enable a feature that isn't meant for SSG 🤔 Please provide a reproduction.

This can be difficult to do, but probably easy to describe:

TL;DR I am parsing request URL pathname and save results into `Astro.locals`, which I then use across Astro components where needed. Works great for static site build time.

In my case, I'm building a static multilingual website and I am using, I guess, my own multilingual solution for that, by only using Middleware, Astro.locals and some JS, getting the locale from URL etc.

I tried to utilize the i18n features at first, but it started to get a bit challenging (also its my first astro project), and at times maybe needless, as I didn't plan to use SSR mode for this project, and while I have some basic i18n configuration in place, it is only used by Astro dev server.

Anyway, I got everything working I wanted, ie. fully localized urls (translated), synched switching of localized content, localized static 404s, hassle-free content management (all content in collections, page navigation and structure generated from content collections), never broken links in menus or in content etc.

**Comment by jurajkapsz** [comment]
> you can't access `Astro.request` for static pages, because there's no request.

That's correct, I don't. I access - have access to - the `Request` object (`Astro.request`?) via Middleware's `context` object while locally running Astro `dev` and Astro `build`.

I can log from [middleware](https://docs.astro.build/en/guides/middleware/#middleware-types) and have it in dev and build terminal logs.

```js
import { defineMiddleware } from "astro:middleware";

export const onRequest = defineMiddleware(({ locals, request }, next) => {
  const { pathname } = new URL(request.url);
  console.log(pathname);
});

// /
// /en/
// /en/articles/
// /en/articles/one/
// /en/articles/two/
// ...
```

I guess, the build process looks at `Astro.request.headers` with the `checkOrigin` config option enabled, and with the combination of `static`/`hybrid` mode, the `headers` are by the looks empty/nonexistent, and it informs about that.

nb I don't access `request.headers` at all, but it's part of the `Request` object, so anyway.

I can make a repro of this if still needed. My english and for sure Astro vocabulary are not the best tbh.

Also, I haven't used the `checkOrigin` option before. I just recently enabled it, preemptively for the Astro Actions I intend to use to process forms. But I have to look it further up if it makes any sense in the first place and/or how to secure my form processing in general. Otherwise, I wouldn't use the `checkOrigin` option, that is, for just `static` mode pages (but now `hybrid` for the Actions).

**Comment by ematipico** [comment]
Thank you, @jurajkapsz, for providing some context.

I am going to close the issue, because you're using `checkOrigin`, which does check `Request` headers, so the warning is expected.

**Comment by jurajkapsz** [comment]
OK, thanks @ematipico , it makes sense in that way, but at the same time it doesn't feel right as I don't implement, use or access anything incorrectly. So its like Astro warns itself but I get to see it, and it clutters the logs quite a bit if there are lots of pages. Finally, it looks like this will be the default behavior in Astro v5, but lets see.

I don't think a repro is useful here anymore, but let me know if anything.

Thanks guys!

**Comment by marceloverdijk** [comment]
@ematipico I'm seeing this message as well using Astro 5 beta 4 for all my pre-rendered pages.

I guess it's related that in Astro 5 `security.checkOrigin` default value was changed to `true`?

Now if some pages are pre-rendered, and some are server-side rendered the default `security.checkOrigin: true` makes sense, but I would say for pre-rendered page the warning does not make sense then, right?

I've many pages so when building I see that warning 1000 times...

**Comment by ematipico** [comment]
@marceloverdijk this issue was fixed yesterday, and it will be released in the next beta version

**Comment by marceloverdijk** [comment]
Thx for confirming 👍
